common/web/gesture-recognizer/src/engine/headless/inputEngineBase.ts

```diff
@@ -20,17 +23,74 @@ interface EventMap<HoveredItemType, StateToken> {
 export abstract class InputEngineBase<HoveredItemType, StateToken = any> extends EventEmitter<EventMap<HoveredItemType, StateToken>> {
   private _activeTouchpoints: GestureSource<HoveredItemType>[] = [];
 
+  // Touch interactions in the browser actually _re-use_ touch IDs once they lapse; the IDs are not lifetime-unique.
+  // This gesture-engine desires lifetime-unique IDs, though, so we map them within this engine to remedy that problem.
+  private readonly identifierMap: Record<number, number> = {};
+  private static IDENTIFIER_SEED = 0;
+
   public stateToken: StateToken;
 
+  protected readonly config: Nonoptional<GestureRecognizerConfiguration<HoveredItemType, StateToken>>;
+
+  public constructor(config: Nonoptional<GestureRecognizerConfiguration<HoveredItemType, StateToken>>) {
+    super();
+    this.config = config;
+  }
+
+  createTouchpoint(identifier: number, isFromTouch: boolean) {
+    // IDs provided to `GestureSource` should be engine-unique.  Unfortunately, the base identifier patterns provided by
+    // browsers don't do this, so we map the browser ID to an engine-internal one.
+    const unique_id = InputEngineBase.IDENTIFIER_SEED++;
+
+    this.identifierMap[identifier] = unique_id;
+
+    const source = new GestureSource<HoveredItemType, StateToken>(unique_id, this.config, isFromTouch);
+    source.stateToken = this.stateToken;
+
+    // Do not add here; it needs special managing for unit tests.
+
+    return source;
+  }
+
+  /**
+   * Calls to this method will cancel any touchpoints whose internal IDs are _not_ included in the parameter.
+   * Designed to facilitate recovery from error cases and peculiar states that sometimes arise when debugging.
+   * @param identifiers
+   */
+  maintainTouchpointsWithIds(identifiers: number[]) {
+    const identifiersToMaintain = identifiers.map((internal_id) => this.identifierMap[internal_id]);
+    const sourcesToDrop: GestureSource<HoveredItemType>[] = [];
+
+    this._activeTouchpoints.forEach((source) => {
+      if(identifiersToMaintain.indexOf(source.rawIdentifier) == -1) {
+        sourcesToDrop.push(source);
+      }
+    });
+
+    sourcesToDrop.forEach((source) => {
+      // Will trigger `.dropTouchpoint` later in the event chain.
+      source.terminate(true);
+    });
```

Comment:
Is this not possible?

```suggestion
    this._activeTouchpoints.forEach(source => {
      if(!identifiersToMaintain.includes(source.rawIdentifier)) {
        // Will trigger `.dropTouchpoint` later in the event chain.
        source.terminate(true);
      }
    });
```

Or, in a more functional approach:

```suggestion
    // Note: terminate will trigger `.dropTouchpoint` later in the event chain.
    this._activeTouchpoints
      .filter(source => !identifiersToMaintain.include(source.rawIdentifier))
      .forEach(source => source.terminate(true));
```